ark/utils/spatial_analysis_utils.py

```diff
@@ -9,30 +9,55 @@
 import os
 
 
-def calc_dist_matrix(label_map, path=None):
+def calc_dist_matrix(label_map, centroid_labels=None, path=None):
     """Generate matrix of distances between center of pairs of cells
 
     Args:
         label_map (np array): array with unique cells given unique pixel labels
+        centroid_labels (dict): the labels for each fov which are
+            needed to add correct coordinates to access distance matrix
+            needs to be of the same length and indexed corresponding to label_map.coords['fovs']
+            if None, then default assume that cell labels are in order for each fov
         path (string): path to save file. If None, then will directly return
     Returns:
         dist_matrix (dict): contains a cells x cells matrix with the euclidian
-            distance between centers of corresponding cells for every fov"""
+            distance between centers of corresponding cells for every fov, note that each distance matrix
+            is of type xarray"""
     # Check that file path exists, if given
+
     if path is not None:
         if not os.path.exists(path):
             raise ValueError("File path not valid")
+
     dist_mats_list = []
+
     # Extract list of fovs
     fovs = list(label_map.coords['fovs'].values)
-    for i in range(0, label_map.shape[0]):
-        props = skimage.measure.regionprops(label_map.loc[fovs[i], :, :, "segmentation_label"].values)
-        a = [props[j].centroid for j in range(len(props))]
-        centroids = np.array(a)
+
+    # generate centroid labels if None
+    if centroid_labels is None:
+        centroid_labels = {}
+
+        for fov_val in fovs:
+            fov_arr = label_map.loc[fov_val, :, :, 'segmentation_label'].values
+            fov_arr_labels = np.unique(fov_arr[fov_arr > 1]).tolist()
+            centroid_labels[fov_val] = fov_arr_labels
```

Comment:
now that I think about it some more, I think we can actually get the correctly indexed labels from props:
```
centroid_labels = np.array([prop.label for prop in props])
```

the comment about filtering out 1's still applies, but we wouldn't have to filter out zeros (since regionprops does this for us).

Reply:
You, my friend, are a lifesaver.

P.S. it turns out the indexing problems I had in the past were because I made a call to np.unique in my random centroid-based generation of a label map. So it wasn't regionprops but rather np.unique that was sorting everything. Big yikes.